<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.11"/>
<title>Augustus: include/contTimeMC.hh Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Augustus
   &#160;<span id="projectnumber">3.3.2</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="dir_d44c64559bbebec7f509842c48db8b23.html">include</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">contTimeMC.hh</div>  </div>
</div><!--header-->
<div class="contents">
<div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;<span class="comment">/*</span></div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;<span class="comment"> * contTimeMC.hh</span></div><div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;<span class="comment"> *</span></div><div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;<span class="comment"> * License: Artistic License, see file LICENSE.TXT or </span></div><div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;<span class="comment"> *          https://opensource.org/licenses/artistic-license-1.0</span></div><div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;</div><div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;<span class="preprocessor">#ifndef _CONTTIMEMC_HH</span></div><div class="line"><a name="l00009"></a><span class="lineno">    9</span>&#160;<span class="preprocessor">#define _CONTTIMEMC_HH</span></div><div class="line"><a name="l00010"></a><span class="lineno">   10</span>&#160;</div><div class="line"><a name="l00011"></a><span class="lineno">   11</span>&#160;<span class="comment">// project includes</span></div><div class="line"><a name="l00012"></a><span class="lineno">   12</span>&#160;<span class="preprocessor">#include &quot;geneticcode.hh&quot;</span></div><div class="line"><a name="l00013"></a><span class="lineno">   13</span>&#160;<span class="preprocessor">#include &quot;matrix.hh&quot;</span></div><div class="line"><a name="l00014"></a><span class="lineno">   14</span>&#160;</div><div class="line"><a name="l00015"></a><span class="lineno">   15</span>&#160;<span class="comment">// standard C/C++ includes</span></div><div class="line"><a name="l00016"></a><span class="lineno">   16</span>&#160;<span class="preprocessor">#include &lt;iostream&gt;</span></div><div class="line"><a name="l00017"></a><span class="lineno">   17</span>&#160;<span class="preprocessor">#include &lt;gsl/gsl_matrix.h&gt;</span></div><div class="line"><a name="l00018"></a><span class="lineno">   18</span>&#160;<span class="preprocessor">#include &lt;gsl/gsl_eigen.h&gt;</span></div><div class="line"><a name="l00019"></a><span class="lineno">   19</span>&#160;<span class="preprocessor">#include &lt;gsl/gsl_linalg.h&gt;</span></div><div class="line"><a name="l00020"></a><span class="lineno">   20</span>&#160;</div><div class="line"><a name="l00021"></a><span class="lineno">   21</span>&#160;<span class="comment">// forward declarations</span></div><div class="line"><a name="l00026"></a><span class="lineno">   26</span>&#160;<span class="comment"></span><span class="keyword">class </span><a class="code" href="classPhyloTree.html">PhyloTree</a>;</div><div class="line"><a name="l00027"></a><span class="lineno">   27</span>&#160;</div><div class="line"><a name="l00028"></a><span class="lineno">   28</span>&#160;<span class="keyword">using namespace </span><a class="code" href="namespacestd.html">std</a>;</div><div class="line"><a name="l00029"></a><span class="lineno">   29</span>&#160;</div><div class="line"><a name="l00036"></a><span class="lineno"><a class="line" href="classEvo.html">   36</a></span>&#160;<span class="keyword">class </span><a class="code" href="classEvo.html">Evo</a> {</div><div class="line"><a name="l00037"></a><span class="lineno">   37</span>&#160;<span class="keyword">public</span>:</div><div class="line"><a name="l00038"></a><span class="lineno">   38</span>&#160;  <a class="code" href="classEvo.html">Evo</a>(<span class="keywordtype">int</span> s) : states(s), m(0), pi(NULL) {};</div><div class="line"><a name="l00039"></a><span class="lineno">   39</span>&#160;    <span class="keyword">virtual</span> ~<a class="code" href="classEvo.html">Evo</a>();</div><div class="line"><a name="l00040"></a><span class="lineno">   40</span>&#160;    <span class="keywordtype">int</span> getNumStates(){<span class="keywordflow">return</span> states;}</div><div class="line"><a name="l00041"></a><span class="lineno">   41</span>&#160;</div><div class="line"><a name="l00042"></a><span class="lineno">   42</span>&#160;    <span class="comment">/* Determine the branch lengths for which matrices P should be stored, at most m.</span></div><div class="line"><a name="l00043"></a><span class="lineno">   43</span>&#160;<span class="comment">     * For trees with few species, say &lt;10, this could be all branch lengths b occuring</span></div><div class="line"><a name="l00044"></a><span class="lineno">   44</span>&#160;<span class="comment">     * in the tree. A value of m=-1 means to store all lengths in b. Memory may be saved</span></div><div class="line"><a name="l00045"></a><span class="lineno">   45</span>&#160;<span class="comment">     * by building m clusters roughly representing all lengths in b.</span></div><div class="line"><a name="l00046"></a><span class="lineno">   46</span>&#160;<span class="comment">     */</span></div><div class="line"><a name="l00047"></a><span class="lineno">   47</span>&#160;    <span class="keywordtype">void</span> setBranchLengths(vector&lt;double&gt; b, <span class="keywordtype">int</span> m=-1);</div><div class="line"><a name="l00048"></a><span class="lineno">   48</span>&#160;    <span class="keywordtype">void</span> printBranchLengths();</div><div class="line"><a name="l00049"></a><span class="lineno">   49</span>&#160;</div><div class="line"><a name="l00050"></a><span class="lineno">   50</span>&#160;    <span class="keyword">virtual</span> <span class="keywordtype">void</span> computeLogPmatrices()=0; <span class="comment">// precomputes and stores the array of matrices</span></div><div class="line"><a name="l00051"></a><span class="lineno">   51</span>&#160;    <span class="keyword">virtual</span> <span class="keywordtype">void</span> addBranchLength(<span class="keywordtype">double</span> b)=0; <span class="comment">//add new branch length b and matrix P(b)</span></div><div class="line"><a name="l00052"></a><span class="lineno">   52</span>&#160;</div><div class="line"><a name="l00053"></a><span class="lineno">   53</span>&#160;    <span class="keywordtype">double</span> getPi(<span class="keywordtype">int</span> i)<span class="keyword"> const </span>{<span class="keywordflow">return</span> pi[i];}  <span class="comment">//equilibrium frequency of state i</span></div><div class="line"><a name="l00054"></a><span class="lineno">   54</span>&#160;    <span class="keywordtype">double</span> getLogPi(<span class="keywordtype">int</span> i)<span class="keyword"> const </span>{<span class="keywordflow">return</span> log(getPi(i));}</div><div class="line"><a name="l00055"></a><span class="lineno">   55</span>&#160;</div><div class="line"><a name="l00056"></a><span class="lineno">   56</span>&#160;    <span class="comment">/*</span></div><div class="line"><a name="l00057"></a><span class="lineno">   57</span>&#160;<span class="comment">     * Returns a pointer to the probability matrix for which t comes closest.</span></div><div class="line"><a name="l00058"></a><span class="lineno">   58</span>&#160;<span class="comment">     * u is the index to a second parameter for which the probability matrices</span></div><div class="line"><a name="l00059"></a><span class="lineno">   59</span>&#160;<span class="comment">     * are stored for different values.</span></div><div class="line"><a name="l00060"></a><span class="lineno">   60</span>&#160;<span class="comment">     * in the case of codon evolution u is the index to omega (dN/ds, selection)</span></div><div class="line"><a name="l00061"></a><span class="lineno">   61</span>&#160;<span class="comment">     * in the case of exon evolution probability matrices currently only</span></div><div class="line"><a name="l00062"></a><span class="lineno">   62</span>&#160;<span class="comment">     * depend on t and thus u is set to 0. It is however conceivable</span></div><div class="line"><a name="l00063"></a><span class="lineno">   63</span>&#160;<span class="comment">     * to allow different rates of exon loss in future.</span></div><div class="line"><a name="l00064"></a><span class="lineno">   64</span>&#160;<span class="comment">     */</span></div><div class="line"><a name="l00065"></a><span class="lineno">   65</span>&#160;    gsl_matrix *getSubMatrixP(<span class="keywordtype">int</span> u, <span class="keywordtype">double</span> t); </div><div class="line"><a name="l00066"></a><span class="lineno">   66</span>&#160;    gsl_matrix *getSubMatrixLogP(<span class="keywordtype">int</span> u, <span class="keywordtype">double</span> t); </div><div class="line"><a name="l00067"></a><span class="lineno">   67</span>&#160;</div><div class="line"><a name="l00068"></a><span class="lineno">   68</span>&#160;    <span class="comment">/*</span></div><div class="line"><a name="l00069"></a><span class="lineno">   69</span>&#160;<span class="comment">     * compute the matrix exponential P(t) = exp(Q*t),</span></div><div class="line"><a name="l00070"></a><span class="lineno">   70</span>&#160;<span class="comment">     * where Q is given by U, diag(lambda) and U^{-1} as computed by eigendecompose</span></div><div class="line"><a name="l00071"></a><span class="lineno">   71</span>&#160;<span class="comment">     */</span></div><div class="line"><a name="l00072"></a><span class="lineno">   72</span>&#160;    gsl_matrix *expQt(<span class="keywordtype">double</span> t, gsl_vector *lambda, gsl_matrix *U, gsl_matrix *Uinv);</div><div class="line"><a name="l00073"></a><span class="lineno">   73</span>&#160;</div><div class="line"><a name="l00074"></a><span class="lineno">   74</span>&#160;<span class="keyword">protected</span>:</div><div class="line"><a name="l00075"></a><span class="lineno">   75</span>&#160;    <span class="keywordtype">int</span> findClosestIndex(vector&lt;double&gt; &amp;v, <span class="keywordtype">double</span> val);</div><div class="line"><a name="l00076"></a><span class="lineno">   76</span>&#160;</div><div class="line"><a name="l00077"></a><span class="lineno">   77</span>&#160;<span class="keyword">protected</span>:</div><div class="line"><a name="l00078"></a><span class="lineno">   78</span>&#160;    <span class="keyword">const</span> <span class="keywordtype">int</span> states; <span class="comment">//number of states (64 in codon model and (currently) 2 in exon model)</span></div><div class="line"><a name="l00079"></a><span class="lineno">   79</span>&#160;    <span class="keywordtype">int</span> m; <span class="comment">// number of branch lengths (times) for which P&#39;s are stored</span></div><div class="line"><a name="l00080"></a><span class="lineno">   80</span>&#160;    <span class="keywordtype">double</span> *pi;</div><div class="line"><a name="l00081"></a><span class="lineno">   81</span>&#160;    vector&lt;double&gt; times; <span class="comment">// sorted vector of branch lengths</span></div><div class="line"><a name="l00082"></a><span class="lineno">   82</span>&#160;    <a class="code" href="classMatrix.html">Matrix&lt;gsl_matrix *&gt;</a> allPs; <span class="comment">// parametrized probability matrices</span></div><div class="line"><a name="l00083"></a><span class="lineno">   83</span>&#160;    <a class="code" href="classMatrix.html">Matrix&lt;gsl_matrix *&gt;</a> allLogPs; <span class="comment">// parametrized log probability matrices</span></div><div class="line"><a name="l00084"></a><span class="lineno">   84</span>&#160;</div><div class="line"><a name="l00085"></a><span class="lineno">   85</span>&#160;};</div><div class="line"><a name="l00086"></a><span class="lineno">   86</span>&#160;</div><div class="line"><a name="l00100"></a><span class="lineno"><a class="line" href="classCodonEvo.html">  100</a></span>&#160;<span class="keyword">class </span><a class="code" href="classCodonEvo.html">CodonEvo</a> : <span class="keyword">public</span> <a class="code" href="classEvo.html">Evo</a> {</div><div class="line"><a name="l00101"></a><span class="lineno">  101</span>&#160;<span class="keyword">public</span>:</div><div class="line"><a name="l00102"></a><span class="lineno">  102</span>&#160;    <a class="code" href="classCodonEvo.html">CodonEvo</a>() : <a class="code" href="classEvo.html">Evo</a>(64), k(0), kappa(4.0) {};</div><div class="line"><a name="l00103"></a><span class="lineno">  103</span>&#160;    ~<a class="code" href="classCodonEvo.html">CodonEvo</a>();</div><div class="line"><a name="l00104"></a><span class="lineno">  104</span>&#160;    <span class="keywordtype">void</span> setKappa(<span class="keywordtype">double</span> kappa){ this-&gt;kappa = kappa;} <span class="comment">// transition/transversion ratio</span></div><div class="line"><a name="l00105"></a><span class="lineno">  105</span>&#160;    <span class="keywordtype">void</span> setPi(<span class="keywordtype">double</span> *pi); <span class="comment">// codon usage</span></div><div class="line"><a name="l00106"></a><span class="lineno">  106</span>&#160;   </div><div class="line"><a name="l00107"></a><span class="lineno">  107</span>&#160;    <span class="comment">/*</span></div><div class="line"><a name="l00108"></a><span class="lineno">  108</span>&#160;<span class="comment">     * Chooses k values for omega around 1, e.g. 0.8, 1, 1.2 for k=3.</span></div><div class="line"><a name="l00109"></a><span class="lineno">  109</span>&#160;<span class="comment">     * Later, one can see which of these values of omega gives the maximum likelihood.</span></div><div class="line"><a name="l00110"></a><span class="lineno">  110</span>&#160;<span class="comment">     */</span></div><div class="line"><a name="l00111"></a><span class="lineno">  111</span>&#160;    <span class="keywordtype">void</span> setOmegas(<span class="keywordtype">int</span> k);</div><div class="line"><a name="l00112"></a><span class="lineno">  112</span>&#160;    <span class="keywordtype">void</span> setPrior(<span class="keywordtype">double</span> sigma = 0.2);</div><div class="line"><a name="l00113"></a><span class="lineno">  113</span>&#160;    <span class="keywordtype">double</span> getPrior(<span class="keywordtype">int</span> u){<span class="keywordflow">return</span> omegaPrior[u];}</div><div class="line"><a name="l00114"></a><span class="lineno">  114</span>&#160;    vector&lt;double&gt; *getOmegas(){<span class="keywordflow">return</span> &amp;omegas;}</div><div class="line"><a name="l00115"></a><span class="lineno">  115</span>&#160;    <span class="keywordtype">double</span> getOmega(<span class="keywordtype">int</span> u){<span class="keywordflow">return</span> omegas[u];}</div><div class="line"><a name="l00116"></a><span class="lineno">  116</span>&#160;    <span class="keywordtype">int</span> getK(){ <span class="keywordflow">return</span> k;}</div><div class="line"><a name="l00117"></a><span class="lineno">  117</span>&#160;    <span class="keywordtype">void</span> printOmegas();</div><div class="line"><a name="l00118"></a><span class="lineno">  118</span>&#160;    <span class="keywordtype">void</span> setAAPostProbs();</div><div class="line"><a name="l00119"></a><span class="lineno">  119</span>&#160;    <span class="keywordtype">void</span> computeLogPmatrices(); <span class="comment">// precomputes and stores the array of matrices</span></div><div class="line"><a name="l00120"></a><span class="lineno">  120</span>&#160;</div><div class="line"><a name="l00121"></a><span class="lineno">  121</span>&#160;    <span class="comment">/*</span></div><div class="line"><a name="l00122"></a><span class="lineno">  122</span>&#160;<span class="comment">     * Returns a pointer to the 64x64 substitution probability matrix</span></div><div class="line"><a name="l00123"></a><span class="lineno">  123</span>&#160;<span class="comment">     * for which t and omega come closest to scored values.</span></div><div class="line"><a name="l00124"></a><span class="lineno">  124</span>&#160;<span class="comment">     */</span></div><div class="line"><a name="l00125"></a><span class="lineno">  125</span>&#160;    gsl_matrix *getSubMatrixLogP(<span class="keywordtype">double</span> omega, <span class="keywordtype">double</span> t);</div><div class="line"><a name="l00126"></a><span class="lineno">  126</span>&#160;    </div><div class="line"><a name="l00127"></a><span class="lineno">  127</span>&#160;    <span class="comment">/*</span></div><div class="line"><a name="l00128"></a><span class="lineno">  128</span>&#160;<span class="comment">     * Computes the substitution probability </span></div><div class="line"><a name="l00129"></a><span class="lineno">  129</span>&#160;<span class="comment">     * P( X(t)=to | X(0)=from, omega), where 1 &lt;= from,to &lt;= 64 are codons</span></div><div class="line"><a name="l00130"></a><span class="lineno">  130</span>&#160;<span class="comment">     * If you want to call this for many values of &#39;from&#39; and &#39;to&#39;, use rather getSubMatrixP above.</span></div><div class="line"><a name="l00131"></a><span class="lineno">  131</span>&#160;<span class="comment">     */</span></div><div class="line"><a name="l00132"></a><span class="lineno">  132</span>&#160;    <span class="keywordtype">double</span> subLogProb(<span class="keywordtype">int</span> from, <span class="keywordtype">int</span> to, <span class="keywordtype">double</span> omega, <span class="keywordtype">double</span> t){</div><div class="line"><a name="l00133"></a><span class="lineno">  133</span>&#160;    gsl_matrix *P = getSubMatrixLogP(omega, t);</div><div class="line"><a name="l00134"></a><span class="lineno">  134</span>&#160;    <span class="keywordflow">return</span> gsl_matrix_get(P, from, to);</div><div class="line"><a name="l00135"></a><span class="lineno">  135</span>&#160;    }</div><div class="line"><a name="l00136"></a><span class="lineno">  136</span>&#160;    <span class="comment">/*</span></div><div class="line"><a name="l00137"></a><span class="lineno">  137</span>&#160;<span class="comment">     * log likelihood of exon candidate pair e1, e1 (required: equal number of codons, gapless)</span></div><div class="line"><a name="l00138"></a><span class="lineno">  138</span>&#160;<span class="comment">     * Used for testing. The pruning algorithm requires tuples of codons instead.</span></div><div class="line"><a name="l00139"></a><span class="lineno">  139</span>&#160;<span class="comment">     */</span></div><div class="line"><a name="l00140"></a><span class="lineno">  140</span>&#160;    <span class="keywordtype">double</span> logLik(<span class="keyword">const</span> <span class="keywordtype">char</span> *e1, <span class="keyword">const</span> <span class="keywordtype">char</span> *e2, <span class="keywordtype">int</span> u, <span class="keywordtype">double</span> t); <span class="comment">// output variables</span></div><div class="line"><a name="l00141"></a><span class="lineno">  141</span>&#160;    <span class="comment">/*</span></div><div class="line"><a name="l00142"></a><span class="lineno">  142</span>&#160;<span class="comment">     * Estimate selection omega on pair of codon sequences.</span></div><div class="line"><a name="l00143"></a><span class="lineno">  143</span>&#160;<span class="comment">     */</span></div><div class="line"><a name="l00144"></a><span class="lineno">  144</span>&#160;    <span class="keywordtype">double</span> estOmegaOnSeqPair(<span class="keyword">const</span> <span class="keywordtype">char</span> *e1, <span class="keyword">const</span> <span class="keywordtype">char</span> *e2, <span class="keywordtype">double</span> t, <span class="comment">// input variables</span></div><div class="line"><a name="l00145"></a><span class="lineno">  145</span>&#160;                 <span class="keywordtype">int</span>&amp; numAliCodons, <span class="keywordtype">int</span> &amp;numSynSubst, <span class="keywordtype">int</span> &amp;numNonSynSubst); <span class="comment">// output variables</span></div><div class="line"><a name="l00146"></a><span class="lineno">  146</span>&#160;    <span class="comment">/*</span></div><div class="line"><a name="l00147"></a><span class="lineno">  147</span>&#160;<span class="comment">     * Estimate omega on a sequence of codon tuples.</span></div><div class="line"><a name="l00148"></a><span class="lineno">  148</span>&#160;<span class="comment">     * only for testing, may need adjustment</span></div><div class="line"><a name="l00149"></a><span class="lineno">  149</span>&#160;<span class="comment">     */</span></div><div class="line"><a name="l00150"></a><span class="lineno">  150</span>&#160;    <span class="keywordtype">double</span> estOmegaOnSeqTuple(vector&lt;string&gt; &amp;seqtuple, <a class="code" href="classPhyloTree.html">PhyloTree</a> *tree,</div><div class="line"><a name="l00151"></a><span class="lineno">  151</span>&#160;                  <span class="keywordtype">int</span> &amp;subst); <span class="comment">// output variables</span></div><div class="line"><a name="l00152"></a><span class="lineno">  152</span>&#160;    <span class="comment">// store log likeliehoods for all omegas for one codon tuple</span></div><div class="line"><a name="l00153"></a><span class="lineno">  153</span>&#160;</div><div class="line"><a name="l00154"></a><span class="lineno">  154</span>&#160;  vector&lt;double&gt; loglikForCodonTuple(vector&lt;string&gt; &amp;seqtuple, <a class="code" href="classPhyloTree.html">PhyloTree</a> *ctree){</div><div class="line"><a name="l00155"></a><span class="lineno">  155</span>&#160;    <span class="keywordtype">int</span> i = 0;</div><div class="line"><a name="l00156"></a><span class="lineno">  156</span>&#160;    <span class="keywordflow">return</span> loglikForCodonTuple(seqtuple, ctree, NULL, i);</div><div class="line"><a name="l00157"></a><span class="lineno">  157</span>&#160;  }</div><div class="line"><a name="l00158"></a><span class="lineno">  158</span>&#160;  vector&lt;double&gt; loglikForCodonTuple(vector&lt;string&gt; &amp;seqtuple, <a class="code" href="classPhyloTree.html">PhyloTree</a> *ctree, <a class="code" href="classPhyloTree.html">PhyloTree</a> *tree, <span class="keywordtype">int</span> &amp;subs);</div><div class="line"><a name="l00159"></a><span class="lineno">  159</span>&#160;    <span class="comment">// for GRK proposal</span></div><div class="line"><a name="l00160"></a><span class="lineno">  160</span>&#160;  <span class="keywordtype">double</span> graphOmegaOnCodonAli(vector&lt;string&gt; &amp;seqtuple, <a class="code" href="classPhyloTree.html">PhyloTree</a> *tree, <span class="keywordtype">int</span> refSpeciesIdx = 0);</div><div class="line"><a name="l00161"></a><span class="lineno">  161</span>&#160;    <span class="comment">/* </span></div><div class="line"><a name="l00162"></a><span class="lineno">  162</span>&#160;<span class="comment">     * add new branch length b</span></div><div class="line"><a name="l00163"></a><span class="lineno">  163</span>&#160;<span class="comment">     * this function is currently not needed, since the pruning algorithm does change the phylogenetic tree</span></div><div class="line"><a name="l00164"></a><span class="lineno">  164</span>&#160;<span class="comment">     */</span></div><div class="line"><a name="l00165"></a><span class="lineno">  165</span>&#160;    <span class="keywordtype">void</span> addBranchLength(<span class="keywordtype">double</span> b){} </div><div class="line"><a name="l00166"></a><span class="lineno">  166</span>&#160;</div><div class="line"><a name="l00167"></a><span class="lineno">  167</span>&#160;<span class="keyword">private</span>:</div><div class="line"><a name="l00168"></a><span class="lineno">  168</span>&#160;    <span class="keywordtype">int</span> k; <span class="comment">// number of different omega values for which P&#39;s are stored</span></div><div class="line"><a name="l00169"></a><span class="lineno">  169</span>&#160;    <span class="keywordtype">double</span> kappa;</div><div class="line"><a name="l00170"></a><span class="lineno">  170</span>&#160;    vector&lt;double&gt; omegas; <span class="comment">// sorted vector of omegas (contains values below, around and above 1)</span></div><div class="line"><a name="l00171"></a><span class="lineno">  171</span>&#160;    vector&lt;double&gt; omegaPrior; <span class="comment">// prior distribution on omega, centered at 1</span></div><div class="line"><a name="l00172"></a><span class="lineno">  172</span>&#160;    vector&lt;double&gt; aaUsage; <span class="comment">// amino acid usage for incorporation into rate matrix</span></div><div class="line"><a name="l00173"></a><span class="lineno">  173</span>&#160;    vector&lt;vector&lt;double&gt; &gt; aaPostProb; <span class="comment">// retreived from BLOSUM (amino acid substitution rate matrix)</span></div><div class="line"><a name="l00174"></a><span class="lineno">  174</span>&#160;};</div><div class="line"><a name="l00175"></a><span class="lineno">  175</span>&#160;</div><div class="line"><a name="l00176"></a><span class="lineno">  176</span>&#160;<span class="comment">/*</span></div><div class="line"><a name="l00177"></a><span class="lineno">  177</span>&#160;<span class="comment"> * 64x64 rate matrix Q for codon substitutions as in Yang, &quot;Computational Molecular Evolution&quot;, (2.7)</span></div><div class="line"><a name="l00178"></a><span class="lineno">  178</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00179"></a><span class="lineno">  179</span>&#160;</div><div class="line"><a name="l00180"></a><span class="lineno">  180</span>&#160;gsl_matrix *getCodonRateMatrix(<span class="keywordtype">double</span> *pi,    <span class="comment">// codon usage, normalized vector with 64 elements</span></div><div class="line"><a name="l00181"></a><span class="lineno">  181</span>&#160;                   <span class="keywordtype">double</span> omega,  <span class="comment">// dN/dS, nonsynonymous/synonymous ratio</span></div><div class="line"><a name="l00182"></a><span class="lineno">  182</span>&#160;                   <span class="keywordtype">double</span> kappa, <span class="comment">// transition/transversion ratio, usually &gt;1</span></div><div class="line"><a name="l00183"></a><span class="lineno">  183</span>&#160;                   vector&lt;vector&lt;double&gt; &gt; *aaPostProb = NULL); <span class="comment">// posterior probs for AA substitutions</span></div><div class="line"><a name="l00184"></a><span class="lineno">  184</span>&#160;</div><div class="line"><a name="l00185"></a><span class="lineno">  185</span>&#160;gsl_matrix *getNonCodingRateMatrix(vector&lt;double&gt; *pi_nuc, <span class="keywordtype">double</span> kappa); <span class="comment">// rate matrix for non-coding model</span></div><div class="line"><a name="l00186"></a><span class="lineno">  186</span>&#160;</div><div class="line"><a name="l00187"></a><span class="lineno">  187</span>&#160;<span class="comment">/*</span></div><div class="line"><a name="l00188"></a><span class="lineno">  188</span>&#160;<span class="comment"> * perform a decompososition of the rate matrix as Q = U * diag(lambda) * U^{-1}</span></div><div class="line"><a name="l00189"></a><span class="lineno">  189</span>&#160;<span class="comment"> * returns 0 if sucessful</span></div><div class="line"><a name="l00190"></a><span class="lineno">  190</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00191"></a><span class="lineno">  191</span>&#160;<span class="keywordtype">int</span> eigendecompose(gsl_matrix *Q,       <span class="comment">// input</span></div><div class="line"><a name="l00192"></a><span class="lineno">  192</span>&#160;           <span class="keywordtype">double</span> *pi,          <span class="comment">// input, must be same as in construction of Q</span></div><div class="line"><a name="l00193"></a><span class="lineno">  193</span>&#160;           gsl_vector *&amp;lambda, <span class="comment">// output, memory for lambda, U and Uinv is allocated within the function</span></div><div class="line"><a name="l00194"></a><span class="lineno">  194</span>&#160;           gsl_matrix *&amp;U,      <span class="comment">// output</span></div><div class="line"><a name="l00195"></a><span class="lineno">  195</span>&#160;           gsl_matrix *&amp;Uinv);  <span class="comment">// output</span></div><div class="line"><a name="l00196"></a><span class="lineno">  196</span>&#160;</div><div class="line"><a name="l00197"></a><span class="lineno">  197</span>&#160;<span class="comment">// eigen decomposition was successful, if Q - U * diag(lambda) * Uinv is close to the 0-matrix</span></div><div class="line"><a name="l00198"></a><span class="lineno">  198</span>&#160;<span class="keywordtype">void</span> validateEigenDecomp(gsl_matrix *Q, gsl_vector *lambda, gsl_matrix *U, gsl_matrix *Uinv, <span class="keywordtype">int</span> states);</div><div class="line"><a name="l00199"></a><span class="lineno">  199</span>&#160;</div><div class="line"><a name="l00200"></a><span class="lineno">  200</span>&#160;<span class="keywordtype">void</span> printCodonMatrix(gsl_matrix *M);</div><div class="line"><a name="l00201"></a><span class="lineno">  201</span>&#160;</div><div class="line"><a name="l00202"></a><span class="lineno">  202</span>&#160;<span class="comment">/*</span></div><div class="line"><a name="l00203"></a><span class="lineno">  203</span>&#160;<span class="comment"> * computes the element-wise natrual logarithm of a matrix P</span></div><div class="line"><a name="l00204"></a><span class="lineno">  204</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00205"></a><span class="lineno">  205</span>&#160;gsl_matrix *log(gsl_matrix *P, <span class="keywordtype">int</span> states);</div><div class="line"><a name="l00206"></a><span class="lineno">  206</span>&#160;</div><div class="line"><a name="l00207"></a><span class="lineno">  207</span>&#160;<span class="comment">/*</span></div><div class="line"><a name="l00208"></a><span class="lineno">  208</span>&#160;<span class="comment"> * @brief class ExonEvo</span></div><div class="line"><a name="l00209"></a><span class="lineno">  209</span>&#160;<span class="comment"> * @details Computes and stores 2x2 probability matrices for exon gain/loss events</span></div><div class="line"><a name="l00210"></a><span class="lineno">  210</span>&#160;<span class="comment"> * P(t) = exp(Q(lambda, mu)*t) for a sample of values for t</span></div><div class="line"><a name="l00211"></a><span class="lineno">  211</span>&#160;<span class="comment"> * and a fixed given value for lambda (rate of exon gain) and mu (rate of exon loss)</span></div><div class="line"><a name="l00212"></a><span class="lineno">  212</span>&#160;<span class="comment"> * Matrices are precomputed and retrieval of P(t) is then a constant time lookup</span></div><div class="line"><a name="l00213"></a><span class="lineno">  213</span>&#160;<span class="comment"> * </span></div><div class="line"><a name="l00214"></a><span class="lineno">  214</span>&#160;<span class="comment"> * @author Mario Stanke</span></div><div class="line"><a name="l00215"></a><span class="lineno">  215</span>&#160;<span class="comment"> * @author Stefanie König</span></div><div class="line"><a name="l00216"></a><span class="lineno">  216</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00217"></a><span class="lineno"><a class="line" href="classExonEvo.html">  217</a></span>&#160;<span class="keyword">class </span><a class="code" href="classExonEvo.html">ExonEvo</a> : <span class="keyword">public</span> <a class="code" href="classEvo.html">Evo</a>{</div><div class="line"><a name="l00218"></a><span class="lineno">  218</span>&#160;</div><div class="line"><a name="l00219"></a><span class="lineno">  219</span>&#160;<span class="keyword">public</span>:</div><div class="line"><a name="l00220"></a><span class="lineno">  220</span>&#160;    <a class="code" href="classExonEvo.html">ExonEvo</a>(<span class="keywordtype">int</span> num_states=2) : <a class="code" href="classEvo.html">Evo</a>(num_states)  {</div><div class="line"><a name="l00221"></a><span class="lineno">  221</span>&#160;    setLambda();</div><div class="line"><a name="l00222"></a><span class="lineno">  222</span>&#160;    setMu();</div><div class="line"><a name="l00223"></a><span class="lineno">  223</span>&#160;    setAliErr();</div><div class="line"><a name="l00224"></a><span class="lineno">  224</span>&#160;    <span class="comment">// mu, lambda and ali_error have to be set before calculating equilibrium frequencies Pi</span></div><div class="line"><a name="l00225"></a><span class="lineno">  225</span>&#160;    setPi();</div><div class="line"><a name="l00226"></a><span class="lineno">  226</span>&#160;    };</div><div class="line"><a name="l00227"></a><span class="lineno">  227</span>&#160;    ~<a class="code" href="classExonEvo.html">ExonEvo</a>(){</div><div class="line"><a name="l00228"></a><span class="lineno">  228</span>&#160;    gsl_matrix_free(U);</div><div class="line"><a name="l00229"></a><span class="lineno">  229</span>&#160;    gsl_matrix_free(Uinv);</div><div class="line"><a name="l00230"></a><span class="lineno">  230</span>&#160;    gsl_vector_free(l);</div><div class="line"><a name="l00231"></a><span class="lineno">  231</span>&#160;    }</div><div class="line"><a name="l00232"></a><span class="lineno">  232</span>&#160;    <span class="keywordtype">void</span> setPi();</div><div class="line"><a name="l00233"></a><span class="lineno">  233</span>&#160;    <span class="keywordtype">void</span> setMu();</div><div class="line"><a name="l00234"></a><span class="lineno">  234</span>&#160;    <span class="keywordtype">void</span> setAliErr();</div><div class="line"><a name="l00235"></a><span class="lineno">  235</span>&#160;    <span class="keywordtype">double</span> getMu()<span class="keyword"> const</span>{<span class="keywordflow">return</span> mu;}</div><div class="line"><a name="l00236"></a><span class="lineno">  236</span>&#160;    <span class="keywordtype">void</span> setLambda();</div><div class="line"><a name="l00237"></a><span class="lineno">  237</span>&#160;    <span class="keywordtype">double</span> getLambda()<span class="keyword"> const</span>{<span class="keywordflow">return</span> lambda;}</div><div class="line"><a name="l00238"></a><span class="lineno">  238</span>&#160;    <span class="keywordtype">double</span> getAliErr()<span class="keyword"> const</span>{<span class="keywordflow">return</span> ali_error;}</div><div class="line"><a name="l00239"></a><span class="lineno">  239</span>&#160;    <span class="keywordtype">void</span> computeLogPmatrices();</div><div class="line"><a name="l00240"></a><span class="lineno">  240</span>&#160;    <span class="keywordtype">void</span> addBranchLength(<span class="keywordtype">double</span> b); <span class="comment">//add new branch length b and matrix P(b)</span></div><div class="line"><a name="l00241"></a><span class="lineno">  241</span>&#160;    <span class="keywordtype">double</span> minBranchLength()<span class="keyword"> const </span>{<span class="keywordflow">return</span> times.front();}</div><div class="line"><a name="l00242"></a><span class="lineno">  242</span>&#160;    <span class="keywordtype">int</span> eigendecompose(gsl_matrix *Q);</div><div class="line"><a name="l00243"></a><span class="lineno">  243</span>&#160;    gsl_matrix *expQt(<span class="keywordtype">double</span> t) {<span class="keywordflow">return</span> Evo::expQt(t,l,U,Uinv);}</div><div class="line"><a name="l00244"></a><span class="lineno">  244</span>&#160;    gsl_matrix *getExonRateMatrix();</div><div class="line"><a name="l00245"></a><span class="lineno">  245</span>&#160;    <span class="keywordtype">void</span> validateEigenDecomp(gsl_matrix *Q){::validateEigenDecomp(Q,l,U,Uinv,states);}</div><div class="line"><a name="l00246"></a><span class="lineno">  246</span>&#160;   </div><div class="line"><a name="l00247"></a><span class="lineno">  247</span>&#160;<span class="keyword">private</span>:</div><div class="line"><a name="l00248"></a><span class="lineno">  248</span>&#160;    <span class="keywordtype">double</span> mu;            <span class="comment">// rate for exon loss</span></div><div class="line"><a name="l00249"></a><span class="lineno">  249</span>&#160;    <span class="keywordtype">double</span> lambda;        <span class="comment">// rate for exon gain</span></div><div class="line"><a name="l00250"></a><span class="lineno">  250</span>&#160;    <span class="keywordtype">double</span> ali_error;     <span class="comment">// rate for alignment error</span></div><div class="line"><a name="l00251"></a><span class="lineno">  251</span>&#160;</div><div class="line"><a name="l00252"></a><span class="lineno">  252</span>&#160;    <span class="comment">// eigen decomposition of exon rate matrix Q = U * diag(l_1,...,l_n) * Uinv</span></div><div class="line"><a name="l00253"></a><span class="lineno">  253</span>&#160;    gsl_matrix *U;</div><div class="line"><a name="l00254"></a><span class="lineno">  254</span>&#160;    gsl_matrix *Uinv;</div><div class="line"><a name="l00255"></a><span class="lineno">  255</span>&#160;    gsl_vector *l;</div><div class="line"><a name="l00256"></a><span class="lineno">  256</span>&#160;};</div><div class="line"><a name="l00257"></a><span class="lineno">  257</span>&#160;</div><div class="line"><a name="l00258"></a><span class="lineno">  258</span>&#160;<span class="comment">/*</span></div><div class="line"><a name="l00259"></a><span class="lineno">  259</span>&#160;<span class="comment"> * @brief class Parsimony</span></div><div class="line"><a name="l00260"></a><span class="lineno">  260</span>&#160;<span class="comment"> * @details transitions from state i to state j are independ of the branch length</span></div><div class="line"><a name="l00261"></a><span class="lineno">  261</span>&#160;<span class="comment"> * and cost -1 if i!=j and 0 if i==j.</span></div><div class="line"><a name="l00262"></a><span class="lineno">  262</span>&#160;<span class="comment"> * can be used to calculate the minimum nunber of codon substitutions</span></div><div class="line"><a name="l00263"></a><span class="lineno">  263</span>&#160;<span class="comment"> * for a given tree topology</span></div><div class="line"><a name="l00264"></a><span class="lineno">  264</span>&#160;<span class="comment"> * </span></div><div class="line"><a name="l00265"></a><span class="lineno">  265</span>&#160;<span class="comment"> * @author Mario Stanke</span></div><div class="line"><a name="l00266"></a><span class="lineno">  266</span>&#160;<span class="comment"> * @author Stefanie König</span></div><div class="line"><a name="l00267"></a><span class="lineno">  267</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00268"></a><span class="lineno"><a class="line" href="classParsimony.html">  268</a></span>&#160;<span class="keyword">class </span><a class="code" href="classParsimony.html">Parsimony</a> : <span class="keyword">public</span> <a class="code" href="classEvo.html">Evo</a>{</div><div class="line"><a name="l00269"></a><span class="lineno">  269</span>&#160;</div><div class="line"><a name="l00270"></a><span class="lineno">  270</span>&#160;<span class="keyword">public</span>:</div><div class="line"><a name="l00271"></a><span class="lineno">  271</span>&#160;    <a class="code" href="classParsimony.html">Parsimony</a>() : <a class="code" href="classEvo.html">Evo</a>(64) {</div><div class="line"><a name="l00272"></a><span class="lineno">  272</span>&#160;    this-&gt;m = 1;</div><div class="line"><a name="l00273"></a><span class="lineno">  273</span>&#160;    this-&gt;times.push_back(1);</div><div class="line"><a name="l00274"></a><span class="lineno">  274</span>&#160;    this-&gt;pi = <span class="keyword">new</span> <span class="keywordtype">double</span>[64];</div><div class="line"><a name="l00275"></a><span class="lineno">  275</span>&#160;    <span class="keywordflow">for</span>(<span class="keywordtype">int</span> i=0; i&lt;64; i++){</div><div class="line"><a name="l00276"></a><span class="lineno">  276</span>&#160;        pi[i]=1;</div><div class="line"><a name="l00277"></a><span class="lineno">  277</span>&#160;    }</div><div class="line"><a name="l00278"></a><span class="lineno">  278</span>&#160;    };</div><div class="line"><a name="l00279"></a><span class="lineno">  279</span>&#160;    <span class="keywordtype">void</span> computeLogPmatrices();</div><div class="line"><a name="l00280"></a><span class="lineno">  280</span>&#160;    <span class="keywordtype">void</span> addBranchLength(<span class="keywordtype">double</span> b){} </div><div class="line"><a name="l00281"></a><span class="lineno">  281</span>&#160;};</div><div class="line"><a name="l00282"></a><span class="lineno">  282</span>&#160;<span class="preprocessor">#endif    // _CONTTIMEMC_HH</span></div><div class="ttc" id="classExonEvo_html"><div class="ttname"><a href="classExonEvo.html">ExonEvo</a></div><div class="ttdef"><b>Definition:</b> contTimeMC.hh:217</div></div>
<div class="ttc" id="namespacestd_html"><div class="ttname"><a href="namespacestd.html">std</a></div></div>
<div class="ttc" id="classParsimony_html"><div class="ttname"><a href="classParsimony.html">Parsimony</a></div><div class="ttdef"><b>Definition:</b> contTimeMC.hh:268</div></div>
<div class="ttc" id="classMatrix_html"><div class="ttname"><a href="classMatrix.html">Matrix&lt; gsl_matrix * &gt;</a></div></div>
<div class="ttc" id="classCodonEvo_html"><div class="ttname"><a href="classCodonEvo.html">CodonEvo</a></div><div class="ttdoc">class CodonEvo </div><div class="ttdef"><b>Definition:</b> contTimeMC.hh:100</div></div>
<div class="ttc" id="classEvo_html"><div class="ttname"><a href="classEvo.html">Evo</a></div><div class="ttdoc">abstract base class for codon and exon evolution </div><div class="ttdef"><b>Definition:</b> contTimeMC.hh:36</div></div>
<div class="ttc" id="classPhyloTree_html"><div class="ttname"><a href="classPhyloTree.html">PhyloTree</a></div><div class="ttdef"><b>Definition:</b> phylotree.hh:106</div></div>
</div><!-- fragment --></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.11
</small></address>
</body>
</html>
