<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.11"/>
<title>Augustus: include/exon_seg.hh Source File</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Augustus
   &#160;<span id="projectnumber">3.3.2</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="dir_d44c64559bbebec7f509842c48db8b23.html">include</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">exon_seg.hh</div>  </div>
</div><!--header-->
<div class="contents">
<div class="fragment"><div class="line"><a name="l00001"></a><span class="lineno">    1</span>&#160;<span class="comment">/*</span></div><div class="line"><a name="l00002"></a><span class="lineno">    2</span>&#160;<span class="comment"> * exon_seg.hh</span></div><div class="line"><a name="l00003"></a><span class="lineno">    3</span>&#160;<span class="comment"> *</span></div><div class="line"><a name="l00004"></a><span class="lineno">    4</span>&#160;<span class="comment"> * License: Artistic License, see file LICENSE.TXT or </span></div><div class="line"><a name="l00005"></a><span class="lineno">    5</span>&#160;<span class="comment"> *          https://opensource.org/licenses/artistic-license-1.0</span></div><div class="line"><a name="l00006"></a><span class="lineno">    6</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00007"></a><span class="lineno">    7</span>&#160;</div><div class="line"><a name="l00008"></a><span class="lineno">    8</span>&#160;<span class="preprocessor">#ifndef _EXON_SEG_HH</span></div><div class="line"><a name="l00009"></a><span class="lineno">    9</span>&#160;<span class="preprocessor">#define _EXON_SEG_HH</span></div><div class="line"><a name="l00010"></a><span class="lineno">   10</span>&#160;</div><div class="line"><a name="l00011"></a><span class="lineno">   11</span>&#160;<span class="preprocessor">#include&lt;iostream&gt;</span></div><div class="line"><a name="l00012"></a><span class="lineno">   12</span>&#160;<span class="preprocessor">#include&lt;fstream&gt;</span></div><div class="line"><a name="l00013"></a><span class="lineno">   13</span>&#160;<span class="preprocessor">#include&lt;math.h&gt;</span></div><div class="line"><a name="l00014"></a><span class="lineno">   14</span>&#160;<span class="preprocessor">#include&lt;getopt.h&gt;</span></div><div class="line"><a name="l00015"></a><span class="lineno">   15</span>&#160;<span class="preprocessor">#include&lt;vector&gt;</span></div><div class="line"><a name="l00016"></a><span class="lineno">   16</span>&#160;<span class="preprocessor">#include&lt;map&gt;</span></div><div class="line"><a name="l00017"></a><span class="lineno">   17</span>&#160;<span class="preprocessor">#include&lt;stdlib.h&gt;</span></div><div class="line"><a name="l00018"></a><span class="lineno">   18</span>&#160;<span class="preprocessor">#include&lt;limits.h&gt;</span></div><div class="line"><a name="l00019"></a><span class="lineno">   19</span>&#160;<span class="keyword">using namespace </span><a class="code" href="namespacestd.html">std</a>;</div><div class="line"><a name="l00020"></a><span class="lineno">   20</span>&#160;</div><div class="line"><a name="l00021"></a><span class="lineno">   21</span>&#160;</div><div class="line"><a name="l00022"></a><span class="lineno">   22</span>&#160;<span class="comment">/* Minus_infinity is a very large number to use as a sign of impossible transitions</span></div><div class="line"><a name="l00023"></a><span class="lineno">   23</span>&#160;<span class="comment"> * in the transition probability matrix</span></div><div class="line"><a name="l00024"></a><span class="lineno">   24</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00025"></a><span class="lineno">   25</span>&#160;<span class="keyword">const</span> <span class="keywordtype">long</span> <span class="keywordtype">int</span> minus_infinity=LONG_MIN;</div><div class="line"><a name="l00026"></a><span class="lineno">   26</span>&#160;<span class="comment">/* maxcov stores the maximum coverage depth that we are considering in storing the distributions</span></div><div class="line"><a name="l00027"></a><span class="lineno">   27</span>&#160;<span class="comment"> * numstates store the number of states </span></div><div class="line"><a name="l00028"></a><span class="lineno">   28</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00029"></a><span class="lineno">   29</span>&#160;<span class="keyword">const</span> <span class="keywordtype">int</span> maxcov =10000;</div><div class="line"><a name="l00030"></a><span class="lineno">   30</span>&#160;<span class="preprocessor">#define NUMSTATES 4</span></div><div class="line"><a name="l00031"></a><span class="lineno">   31</span>&#160;<span class="comment">/* EXONP denotes positive exon EXONM denotes the negative exon INTRONJ and INTRONI denote the </span></div><div class="line"><a name="l00032"></a><span class="lineno">   32</span>&#160;<span class="comment"> * J and I labels of introns Together these four are the states that we are considering</span></div><div class="line"><a name="l00033"></a><span class="lineno">   33</span>&#160;<span class="comment"> * in our HMM</span></div><div class="line"><a name="l00034"></a><span class="lineno">   34</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00035"></a><span class="lineno">   35</span>&#160;<span class="keyword">const</span> <span class="keywordtype">int</span> EXONP =0;</div><div class="line"><a name="l00036"></a><span class="lineno">   36</span>&#160;<span class="keyword">const</span> <span class="keywordtype">int</span> EXONM =1;</div><div class="line"><a name="l00037"></a><span class="lineno">   37</span>&#160;<span class="keyword">const</span> <span class="keywordtype">int</span> INTRONI =2;</div><div class="line"><a name="l00038"></a><span class="lineno">   38</span>&#160;<span class="keyword">const</span> <span class="keywordtype">int</span> INTRONJ =3;</div><div class="line"><a name="l00039"></a><span class="lineno">   39</span>&#160;<span class="comment">/* STRANDP STRANDM and STRANDB denote the positive negative and unknown strand respectively</span></div><div class="line"><a name="l00040"></a><span class="lineno">   40</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00041"></a><span class="lineno">   41</span>&#160;<span class="keyword">const</span> <span class="keywordtype">int</span> STRANDP =0;</div><div class="line"><a name="l00042"></a><span class="lineno">   42</span>&#160;<span class="keyword">const</span> <span class="keywordtype">int</span> STRANDM =1;</div><div class="line"><a name="l00043"></a><span class="lineno">   43</span>&#160;<span class="keyword">const</span> <span class="keywordtype">int</span> STRANDB =2;</div><div class="line"><a name="l00044"></a><span class="lineno">   44</span>&#160;<span class="comment">/* pott_gamma denotes the gamma value that we will be using in calculating the</span></div><div class="line"><a name="l00045"></a><span class="lineno">   45</span>&#160;<span class="comment"> * Pott&#39;s functionals</span></div><div class="line"><a name="l00046"></a><span class="lineno">   46</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00047"></a><span class="lineno">   47</span>&#160;<span class="keyword">const</span> <span class="keywordtype">int</span> pott_gamma =1500;</div><div class="line"><a name="l00048"></a><span class="lineno">   48</span>&#160;<span class="comment">/* moving_window stores the window size that we will be using while finding the </span></div><div class="line"><a name="l00049"></a><span class="lineno">   49</span>&#160;<span class="comment"> * splice sites using the template matching technique</span></div><div class="line"><a name="l00050"></a><span class="lineno">   50</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00051"></a><span class="lineno">   51</span>&#160;<span class="keyword">const</span> <span class="keywordtype">int</span> moving_window=70;</div><div class="line"><a name="l00052"></a><span class="lineno">   52</span>&#160;<span class="comment">/* The convergence limit that we use for calculating the lambda values using the iterative technique</span></div><div class="line"><a name="l00053"></a><span class="lineno">   53</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00054"></a><span class="lineno">   54</span>&#160;<span class="keyword">const</span> <span class="keywordtype">double</span> convergence_limit=0.01;</div><div class="line"><a name="l00055"></a><span class="lineno">   55</span>&#160;<span class="comment">/* L stores the threshold between introns and exons which we use when we try to estimate the</span></div><div class="line"><a name="l00056"></a><span class="lineno">   56</span>&#160;<span class="comment"> * the distribution using the train_function which calculates the distribution using the exons</span></div><div class="line"><a name="l00057"></a><span class="lineno">   57</span>&#160;<span class="comment"> * predicted from a gff file</span></div><div class="line"><a name="l00058"></a><span class="lineno">   58</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00059"></a><span class="lineno">   59</span>&#160;<span class="keyword">const</span> <span class="keywordtype">int</span> L=4;</div><div class="line"><a name="l00060"></a><span class="lineno">   60</span>&#160;</div><div class="line"><a name="l00061"></a><span class="lineno">   61</span>&#160;<span class="comment">/* lengths stores the length for the states </span></div><div class="line"><a name="l00062"></a><span class="lineno">   62</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00063"></a><span class="lineno">   63</span>&#160;<span class="keyword">const</span> <span class="keywordtype">int</span> lengths[]={1,1,38,1};</div><div class="line"><a name="l00064"></a><span class="lineno">   64</span>&#160;</div><div class="line"><a name="l00065"></a><span class="lineno">   65</span>&#160;<span class="comment">/* fragment is a structure which stores small streches of the hints generated in the first part and also</span></div><div class="line"><a name="l00066"></a><span class="lineno">   66</span>&#160;<span class="comment"> * the necessary information associated with them</span></div><div class="line"><a name="l00067"></a><span class="lineno">   67</span>&#160;<span class="comment"> */</span></div><div class="line"><a name="l00068"></a><span class="lineno">   68</span>&#160;<span class="keyword">typedef</span> <span class="keyword">struct </span><a class="code" href="classfragment.html">fragment</a>{</div><div class="line"><a name="l00069"></a><span class="lineno">   69</span>&#160;  <span class="keywordtype">int</span> state,start,end;</div><div class="line"><a name="l00070"></a><span class="lineno">   70</span>&#160;  <span class="keywordtype">double</span> avg_cov;</div><div class="line"><a name="l00071"></a><span class="lineno">   71</span>&#160;  <span class="comment">/* cov_exp stores the coverage depth for that stretch</span></div><div class="line"><a name="l00072"></a><span class="lineno">   72</span>&#160;<span class="comment">   * in all the experiments</span></div><div class="line"><a name="l00073"></a><span class="lineno">   73</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00074"></a><span class="lineno">   74</span>&#160;  vector&lt;double&gt; cov_exp;</div><div class="line"><a name="l00075"></a><span class="lineno">   75</span>&#160;} <a class="code" href="classfragment.html">fragment</a>;</div><div class="line"><a name="l00076"></a><span class="lineno">   76</span>&#160;</div><div class="line"><a name="l00077"></a><span class="lineno">   77</span>&#160;</div><div class="line"><a name="l00081"></a><span class="lineno"><a class="line" href="classdataset.html">   81</a></span>&#160;<span class="keyword">class </span><a class="code" href="classdataset.html">dataset</a>{</div><div class="line"><a name="l00082"></a><span class="lineno">   82</span>&#160;  </div><div class="line"><a name="l00083"></a><span class="lineno">   83</span>&#160;<span class="keyword">public</span>:</div><div class="line"><a name="l00084"></a><span class="lineno">   84</span>&#160;  <span class="comment">/* define the map used to store the data</span></div><div class="line"><a name="l00085"></a><span class="lineno">   85</span>&#160;<span class="comment">   * The top most vector stores different strands </span></div><div class="line"><a name="l00086"></a><span class="lineno">   86</span>&#160;<span class="comment">   * the next one stores the different tracks</span></div><div class="line"><a name="l00087"></a><span class="lineno">   87</span>&#160;<span class="comment">   * The next vector contains the score at each position in the track</span></div><div class="line"><a name="l00088"></a><span class="lineno">   88</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00089"></a><span class="lineno">   89</span>&#160;  map&lt;string, vector&lt; vector&lt; vector&lt;int&gt; &gt; &gt; &gt; database;</div><div class="line"><a name="l00090"></a><span class="lineno">   90</span>&#160;  <span class="comment">/* define the iterator to loop through the map</span></div><div class="line"><a name="l00091"></a><span class="lineno">   91</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00092"></a><span class="lineno">   92</span>&#160;  map&lt;string, vector&lt; vector&lt; vector&lt;int&gt; &gt; &gt; &gt;::iterator it;</div><div class="line"><a name="l00093"></a><span class="lineno">   93</span>&#160;  <span class="keywordtype">int</span> max_depth;</div><div class="line"><a name="l00094"></a><span class="lineno">   94</span>&#160;  <span class="keywordtype">int</span> max_len;</div><div class="line"><a name="l00095"></a><span class="lineno">   95</span>&#160;  <span class="keywordtype">int</span> no_of_tracks;</div><div class="line"><a name="l00096"></a><span class="lineno">   96</span>&#160;  vector&lt;string&gt; name_of_tracks;</div><div class="line"><a name="l00097"></a><span class="lineno">   97</span>&#160;  <span class="keywordtype">double</span> avg_depth;</div><div class="line"><a name="l00098"></a><span class="lineno">   98</span>&#160;} ;</div><div class="line"><a name="l00099"></a><span class="lineno">   99</span>&#160;</div><div class="line"><a name="l00100"></a><span class="lineno"><a class="line" href="classexon__segmentation.html">  100</a></span>&#160;<span class="keyword">class </span><a class="code" href="classexon__segmentation.html">exon_segmentation</a>{</div><div class="line"><a name="l00101"></a><span class="lineno">  101</span>&#160;<span class="comment">/* Variable lambda1 stores the parameter for the distribution of intron</span></div><div class="line"><a name="l00102"></a><span class="lineno">  102</span>&#160;<span class="comment">   * and lambda2 stores the parameter for the distribution of exons</span></div><div class="line"><a name="l00103"></a><span class="lineno">  103</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00104"></a><span class="lineno">  104</span>&#160;  <span class="keywordtype">double</span> lambda1;</div><div class="line"><a name="l00105"></a><span class="lineno">  105</span>&#160;  <span class="keywordtype">double</span> lambda2;</div><div class="line"><a name="l00106"></a><span class="lineno">  106</span>&#160;  <span class="comment">/* stores the probability distribution for each experiment</span></div><div class="line"><a name="l00107"></a><span class="lineno">  107</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00108"></a><span class="lineno">  108</span>&#160;  vector&lt;vector&lt;double&gt; &gt; prob_dist_exon;</div><div class="line"><a name="l00109"></a><span class="lineno">  109</span>&#160;  vector&lt;vector&lt;double&gt; &gt; prob_dist_intron;</div><div class="line"><a name="l00110"></a><span class="lineno">  110</span>&#160;</div><div class="line"><a name="l00111"></a><span class="lineno">  111</span>&#160;</div><div class="line"><a name="l00112"></a><span class="lineno">  112</span>&#160;<span class="keyword">public</span>:</div><div class="line"><a name="l00113"></a><span class="lineno">  113</span>&#160;  <span class="comment">//constructor</span></div><div class="line"><a name="l00114"></a><span class="lineno">  114</span>&#160;  <a class="code" href="classexon__segmentation.html">exon_segmentation</a>();</div><div class="line"><a name="l00115"></a><span class="lineno">  115</span>&#160;  <span class="comment">/* function to calculate the geometric distribution parameters (lambdas)</span></div><div class="line"><a name="l00116"></a><span class="lineno">  116</span>&#160;<span class="comment">   * which will be used later to calculate the distributions. </span></div><div class="line"><a name="l00117"></a><span class="lineno">  117</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00118"></a><span class="lineno">  118</span>&#160;  <span class="keywordtype">void</span> calculate_lambda(<a class="code" href="classdataset.html">dataset</a> &amp;coverage_info, <span class="keywordtype">int</span> c,<span class="keywordtype">float</span> s);</div><div class="line"><a name="l00119"></a><span class="lineno">  119</span>&#160;  <span class="comment">/* calculates the geometric distribution values using the lambda values and stores them</span></div><div class="line"><a name="l00120"></a><span class="lineno">  120</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00121"></a><span class="lineno">  121</span>&#160;  <span class="keywordtype">void</span> prob_density(<span class="keywordtype">int</span> max,<span class="keywordtype">int</span> no_of_tracks);</div><div class="line"><a name="l00122"></a><span class="lineno">  122</span>&#160;  <span class="comment">/* this function calculates the probability distribution values using an input gff file for training</span></div><div class="line"><a name="l00123"></a><span class="lineno">  123</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00124"></a><span class="lineno">  124</span>&#160;  <span class="keywordtype">void</span> train_function(<a class="code" href="classdataset.html">dataset</a> &amp;coverage_info,<span class="keywordtype">string</span> inputfile,<span class="keywordtype">string</span> key);</div><div class="line"><a name="l00125"></a><span class="lineno">  125</span>&#160;  <span class="comment">/* functions which return the distribution value for a particular coverage depth</span></div><div class="line"><a name="l00126"></a><span class="lineno">  126</span>&#160;<span class="comment">   * by accessing the corresponding index in the distribution table</span></div><div class="line"><a name="l00127"></a><span class="lineno">  127</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00128"></a><span class="lineno">  128</span>&#160;  <span class="keywordtype">double</span> prob_value_exon(<span class="keywordtype">int</span> num,<span class="keywordtype">int</span> track);</div><div class="line"><a name="l00129"></a><span class="lineno">  129</span>&#160;  <span class="keywordtype">double</span> prob_value_intron(<span class="keywordtype">int</span> num,<span class="keywordtype">int</span> track);</div><div class="line"><a name="l00130"></a><span class="lineno">  130</span>&#160;  <span class="comment">/* Find the splice sites</span></div><div class="line"><a name="l00131"></a><span class="lineno">  131</span>&#160;<span class="comment">   * INCOMPLETE</span></div><div class="line"><a name="l00132"></a><span class="lineno">  132</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00133"></a><span class="lineno">  133</span>&#160;  <span class="keywordtype">void</span> find_splice(vector&lt; vector&lt; vector&lt;int&gt; &gt; &gt; &amp;input_set,<a class="code" href="classdataset.html">dataset</a> &amp;coverage_info);</div><div class="line"><a name="l00134"></a><span class="lineno">  134</span>&#160;  <span class="comment">/* reads the file and stores it in the the object of the class dataset</span></div><div class="line"><a name="l00135"></a><span class="lineno">  135</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00136"></a><span class="lineno">  136</span>&#160;  <span class="keywordtype">void</span> read_file(<a class="code" href="classdataset.html">dataset</a> &amp;coverage_info,<span class="keywordtype">string</span> filename);</div><div class="line"><a name="l00137"></a><span class="lineno">  137</span>&#160;  <span class="comment">/* it takes as input pointer to a 2-D vector, </span></div><div class="line"><a name="l00138"></a><span class="lineno">  138</span>&#160;<span class="comment">   * the input dataset for a chromosome and</span></div><div class="line"><a name="l00139"></a><span class="lineno">  139</span>&#160;<span class="comment">   * stores the emission probability values in it</span></div><div class="line"><a name="l00140"></a><span class="lineno">  140</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00141"></a><span class="lineno">  141</span>&#160;  <span class="keywordtype">void</span> calculate_emissions( vector&lt; vector&lt;double&gt; &gt; &amp;emission_probs,<span class="keywordtype">int</span> max_len,vector&lt; vector&lt; vector&lt;int&gt; &gt; &gt; &amp;input_set,<span class="keywordtype">int</span> chunksize,<span class="keywordtype">int</span> iteration);</div><div class="line"><a name="l00142"></a><span class="lineno">  142</span>&#160;  <span class="comment">/* takes as input a row and column and calculates the value to be filled </span></div><div class="line"><a name="l00143"></a><span class="lineno">  143</span>&#160;<span class="comment">   * in that element in the gamma matrix</span></div><div class="line"><a name="l00144"></a><span class="lineno">  144</span>&#160;<span class="comment">   * there is also a pointer spred to store the predecessor state in it</span></div><div class="line"><a name="l00145"></a><span class="lineno">  145</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00146"></a><span class="lineno">  146</span>&#160;  <span class="keywordtype">void</span> fill_viterbi(vector&lt; vector&lt;double&gt; &gt; &amp;gamma, <span class="keywordtype">int</span> numstates,<span class="keywordtype">int</span> j, <span class="keywordtype">int</span> s,<span class="keywordtype">double</span> emission_val,<span class="keywordtype">double</span> a[][NUMSTATES],<span class="keywordtype">int</span>* spred);</div><div class="line"><a name="l00147"></a><span class="lineno">  147</span>&#160;  <span class="comment">/* this function implements the viterbi algorithm</span></div><div class="line"><a name="l00148"></a><span class="lineno">  148</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00149"></a><span class="lineno">  149</span>&#160;  vector&lt;int&gt; viterbi(vector&lt; vector&lt; vector&lt;int&gt; &gt; &gt; &amp;input_set,<span class="keywordtype">int</span> max_score,<span class="keywordtype">int</span> max_len,<span class="keywordtype">int</span> chunksize,<span class="keywordtype">int</span> iteration);</div><div class="line"><a name="l00150"></a><span class="lineno">  150</span>&#160;  <span class="comment">/* function to convert the state sequence into segments and also to store the avg coverage depth</span></div><div class="line"><a name="l00151"></a><span class="lineno">  151</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00152"></a><span class="lineno">  152</span>&#160;  vector&lt; fragment &gt; segment( vector&lt;int&gt; state_seq,vector&lt; vector&lt; vector&lt;int&gt; &gt; &gt; &amp;input_set,<span class="keywordtype">int</span> include_intron,<span class="keywordtype">int</span> no_of_tracks );</div><div class="line"><a name="l00153"></a><span class="lineno">  153</span>&#160;  <span class="comment">/* This function takes input the average coverage depth and transforms them</span></div><div class="line"><a name="l00154"></a><span class="lineno">  154</span>&#160;<span class="comment">   * into some usable form in the pott&#39;s functional</span></div><div class="line"><a name="l00155"></a><span class="lineno">  155</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00156"></a><span class="lineno">  156</span>&#160;  <span class="keywordtype">double</span> pott_convert(<span class="keywordtype">double</span> d);</div><div class="line"><a name="l00157"></a><span class="lineno">  157</span>&#160;  <span class="comment">/* calculate m[a,b]=(psum[b]-psum[a-1])/(psuml[b]-psuml[a-1])</span></div><div class="line"><a name="l00158"></a><span class="lineno">  158</span>&#160;<span class="comment">   * to be used for calculating pott&#39;s functionals</span></div><div class="line"><a name="l00159"></a><span class="lineno">  159</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00160"></a><span class="lineno">  160</span>&#160;  <span class="keywordtype">double</span> calculate_m(<span class="keywordtype">int</span> a,<span class="keywordtype">int</span> b,vector&lt;vector&lt;int&gt; &gt; &amp;psuml,vector&lt;vector&lt;double&gt; &gt; &amp;psum,<span class="keywordtype">int</span> exp);</div><div class="line"><a name="l00161"></a><span class="lineno">  161</span>&#160;  <span class="comment">/* Calculate S(a,b) to be used in the pott&#39;s functional</span></div><div class="line"><a name="l00162"></a><span class="lineno">  162</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00163"></a><span class="lineno">  163</span>&#160;  <span class="keywordtype">double</span> calculate_s(<span class="keywordtype">int</span> a,<span class="keywordtype">int</span> b,vector&lt;vector&lt;int&gt; &gt; &amp;psuml,vector&lt;vector&lt;double&gt; &gt; &amp;psum,vector&lt;vector&lt;double&gt; &gt; &amp;psqrsum,vector&lt;vector&lt;double&gt; &gt; &amp;pcubesum,vector&lt;fragment&gt; &amp;segments,<span class="keywordtype">int</span> exp);</div><div class="line"><a name="l00164"></a><span class="lineno">  164</span>&#160;  <span class="comment">/* the function which takes as input the vector of fragments and clusters the exons together to form genes</span></div><div class="line"><a name="l00165"></a><span class="lineno">  165</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00166"></a><span class="lineno">  166</span>&#160;  vector&lt;int&gt; cluster_exons(vector&lt;fragment&gt; segments,<span class="keywordtype">int</span> no_of_tracks);</div><div class="line"><a name="l00167"></a><span class="lineno">  167</span>&#160;</div><div class="line"><a name="l00168"></a><span class="lineno">  168</span>&#160;  <span class="comment">/* Takes the dataset, the name of output file and the threshold c and the ratio of complement </span></div><div class="line"><a name="l00169"></a><span class="lineno">  169</span>&#160;<span class="comment">   * of the parameters(s) as input and makes the gff file</span></div><div class="line"><a name="l00170"></a><span class="lineno">  170</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00171"></a><span class="lineno">  171</span>&#160;  <span class="keywordtype">void</span> make_gff(<a class="code" href="classdataset.html">dataset</a> &amp;coverage_info,<span class="keywordtype">string</span> outputfile,<span class="keywordtype">int</span> c,<span class="keywordtype">float</span> s,<span class="keywordtype">string</span> train_file,<span class="keywordtype">int</span> flag_r,<span class="keywordtype">string</span> bed_file,<span class="keywordtype">string</span> augustus_hints,<span class="keywordtype">int</span> chunksize);</div><div class="line"><a name="l00172"></a><span class="lineno">  172</span>&#160;  <span class="comment">/* takes the dataset and vector of fragments as input and</span></div><div class="line"><a name="l00173"></a><span class="lineno">  173</span>&#160;<span class="comment">   * stores the distribution of number of positions with the </span></div><div class="line"><a name="l00174"></a><span class="lineno">  174</span>&#160;<span class="comment">   * coverage depths for both introns and exons </span></div><div class="line"><a name="l00175"></a><span class="lineno">  175</span>&#160;<span class="comment">   * in separate files for plotting</span></div><div class="line"><a name="l00176"></a><span class="lineno">  176</span>&#160;<span class="comment">   */</span></div><div class="line"><a name="l00177"></a><span class="lineno">  177</span>&#160;  <span class="keywordtype">void</span> make_graph(<a class="code" href="classdataset.html">dataset</a> &amp;coverage_info,vector&lt;fragment &gt; &amp;segments);</div><div class="line"><a name="l00178"></a><span class="lineno">  178</span>&#160;};</div><div class="line"><a name="l00179"></a><span class="lineno">  179</span>&#160;</div><div class="line"><a name="l00180"></a><span class="lineno">  180</span>&#160;<span class="preprocessor">#endif</span></div><div class="ttc" id="namespacestd_html"><div class="ttname"><a href="namespacestd.html">std</a></div></div>
<div class="ttc" id="classdataset_html"><div class="ttname"><a href="classdataset.html">dataset</a></div><div class="ttdoc">It is a small class to store the information extracted from the wig file. </div><div class="ttdef"><b>Definition:</b> exon_seg.hh:81</div></div>
<div class="ttc" id="classexon__segmentation_html"><div class="ttname"><a href="classexon__segmentation.html">exon_segmentation</a></div><div class="ttdef"><b>Definition:</b> exon_seg.hh:100</div></div>
<div class="ttc" id="classfragment_html"><div class="ttname"><a href="classfragment.html">fragment</a></div><div class="ttdoc">gapless alignment fragment </div><div class="ttdef"><b>Definition:</b> alignment.hh:24</div></div>
</div><!-- fragment --></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.11
</small></address>
</body>
</html>
